File: r_sql/sql.R
#packages
library(DBI)
library(dplyr)


#creating database
if (file.exists("airline2.db")) 
  file.remove("airline2.db")

conn <- dbConnect(RSQLite::SQLite(), "airline2.db")

#adding the tables
ontime_2000 <- read.csv("../files/2000.csv", header = TRUE)
ontime_2001 <- read.csv("../files/2001.csv", header = TRUE)
ontime_2002 <- read.csv("../files/2002.csv", header = TRUE)
ontime_2003 <- read.csv("../files/2003.csv", header = TRUE)
ontime_2004 <- read.csv("../files/2004.csv", header = TRUE)
ontime_2005 <- read.csv("../files/2005.csv", header = TRUE)
airports <- read.csv("../files/airports.csv", header = TRUE)
carriers <- read.csv("../files/carriers.csv", header = TRUE)
planes <- read.csv("../files/plane-data.csv", header = TRUE)
combine_ontimes <- bind_rows(ontime_2000, ontime_2001, ontime_2002, ontime_2003, ontime_2004, ontime_2005)
dbWriteTable(conn, "ontime", combine_ontimes)
dbWriteTable(conn, "airports", airports)
dbWriteTable(conn, "carriers", carriers)
dbWriteTable(conn, "planes", planes)

#creating reference to tables for dyplr
ontime_db <- tbl(conn, "ontime")
airports_db <- tbl(conn, "airports")
carriers_db <- tbl(conn, "carriers")
planes_db <- tbl(conn, "planes")

#Question 2
#Which of the following airplanes has the 
#lowest associated average departure delay (excluding cancelled and diverted flights)?

#DBI
q2_DBI <- dbGetQuery(conn,
                      "SELECT model, avg(DepDelay) as average_delay
                      FROM ontime JOIN planes USING (TailNum)
                      WHERE ontime.diverted = 0 AND ontime.cancelled = 0 AND model IN ('737-230', 'ERJ 190-100 IGW', 'A330-223', '737-282')
                      GROUP BY model
                      ORDER BY average_delay")
q2_DBI

#Dyplr
q2_Dyplr <- inner_join(ontime_db, planes_db, by = c("TailNum" = "tailnum")) %>% 
  filter(Diverted == 0, Cancelled == 0, model %in% c('737-230', 'ERJ 190-100 IGW', 'A330-223', '737-282')) %>% 
  group_by(model) %>% 
  summarize(average_delay = mean(DepDelay, na.rm = TRUE)) %>% 
  arrange(average_delay)

q2_Dyplr

#answer: 737-282

#-------

#Question 3
#Which of the following cities has the highest number of inbound flights
#(excluding cancelled flights)

#DBI
q3_DBI <- dbGetQuery(conn,
                      "SELECT airports.city AS city, COUNT(*) AS count_flights
                      FROM ontime JOIN airports ON ontime.dest = airports.iata
                      WHERE ontime.cancelled = 0 AND city IN ('Chicago', 'Atlanta', 'New York', 'Houston')
                      GROUP BY city
                      ORDER BY count_flights DESC")
q3_DBI

#Dyplr
q3_Dyplr <- inner_join(ontime_db, airports_db, by = c("Dest" = "iata")) %>% 
  filter(Cancelled == 0, city %in% c('Chicago', 'Atlanta', 'New York', 'Houston')) %>% 
  group_by(city) %>% 
  summarize(count_flights = n()) %>% 
  arrange(desc(count_flights))

q3_Dyplr

#answer: Chicago
#doubt: Warning message: Closing open result set, pending rows 

#-------

#Question 4
#Which of the following companies has the highest number of cancelled flights?

#DBI
q4_DBI <- dbGetQuery(conn,
                      "SELECT Description, sum(cancelled) as number_cancelled_flights
                      FROM ontime JOIN carriers ON ontime.UniqueCarrier = carriers.Code
                      WHERE Description IN ('United Air Lines Inc.', 'American Airlines Inc.', 'Pinnacle Airlines Inc.', 'Delta Air Lines Inc.')
                      GROUP BY Description
                      ORDER BY number_cancelled_flights DESC")
q4_DBI

#Dyplr
q4_Dyplr <- inner_join(ontime_db, carriers_db, by = c("UniqueCarrier" = "Code")) %>% 
  filter(Description %in% c('United Air Lines Inc.', 'American Airlines Inc.', 'Pinnacle Airlines Inc.', 'Delta Air Lines Inc.')) %>% 
  group_by(Description) %>% 
  summarize(number_cancelled_flights = sum(Cancelled, na.rm = TRUE)) %>% 
  arrange(desc(number_cancelled_flights))

q4_Dyplr

#answer: Delta Air Lines Inc.
#doubt: i lost pinnacle?

#-------

#Question 5
#Which of the following companies has the highest number of cancelled flights,
#relative to their number of total flights?

#DBI
q5_DBI <- dbGetQuery(conn,
                      "SELECT Description, (sum(cancelled)*1.0/count(*)) as cancellation_rate
                      FROM ontime JOIN carriers ON ontime.UniqueCarrier = carriers.Code
                      WHERE Description IN ('United Air Lines Inc.', 'American Airlines Inc.', 'Pinnacle Airlines Inc.', 'Delta Air Lines Inc.')
                      GROUP BY Description
                      HAVING count(*)>0
                      ORDER BY cancellation_rate DESC")
q5_DBI

#Dyplr
q5_Dyplr <- inner_join(ontime_db, carriers_db, by = c("UniqueCarrier" = "Code")) %>% 
  filter(Description %in% c('United Air Lines Inc.', 'American Airlines Inc.', 'Pinnacle Airlines Inc.', 'Delta Air Lines Inc.')) %>% 
  group_by(Description) %>% 
  summarize(cancellation_rate = sum(Cancelled, na.rm = TRUE) * 1.0 / n()) %>%
  filter(n()>0) %>% 
  arrange(desc(cancellation_rate))

q5_Dyplr


#answer: United Air Lines Inc.
#doubt: i lost pinnacle?

#answers in csv

#DBI

write.csv(q2_DBI, "q2_DBI.csv", row.names = FALSE)
write.csv(q3_DBI, "q3_DBI.csv", row.names = FALSE)
write.csv(q4_DBI, "q4_DBI.csv", row.names = FALSE)
write.csv(q5_DBI, "q5_DBI.csv", row.names = FALSE)

#Dyplr
write.csv(q2_Dyplr, "q2_Dyplr.csv", row.names = FALSE)
write.csv(q3_Dyplr, "q3_Dyplr.csv", row.names = FALSE)
write.csv(q4_Dyplr, "q4_Dyplr.csv", row.names = FALSE)
write.csv(q5_Dyplr, "q5_Dyplr.csv", row.names = FALSE)

#Disconnect and clean environment
dbDisconnect(conn)
rm(list = ls())

File: python_sql/sql.py
# -*- coding: utf-8 -*-
"""
Created on Mon Mar  3 19:22:50 2025

@author: joana
"""

#libraries
import os
import sqlite3
import pandas as pd

#creating database
try:
    os.remove('airline2.db')
except OSError:
    pass

conn = sqlite3.connect('airline2.db')


#adding tables
ontime_2000 = pd.read_csv("../files/2000.csv")
ontime_2001 = pd.read_csv("../files/2001.csv", encoding = 'ISO-8859-2')
ontime_2002 = pd.read_csv("../files/2002.csv", encoding = 'ISO-8859-2')
ontime_2003 = pd.read_csv("../files/2003.csv")
ontime_2004 = pd.read_csv("../files/2004.csv")
ontime_2005 = pd.read_csv("../files/2005.csv")
airports = pd.read_csv("../files/airports.csv")
carriers = pd.read_csv("../files/carriers.csv")
planes = pd.read_csv("../files/plane-data.csv")


ontime_2000.to_sql('ontime', con = conn, index = False, if_exists = 'append', chunksize = 100000)
ontime_2001.to_sql('ontime', con = conn, index = False, if_exists = 'append', chunksize = 100000)
ontime_2002.to_sql('ontime', con = conn, index = False, if_exists = 'append', chunksize = 100000)
ontime_2003.to_sql('ontime', con = conn, index = False, if_exists = 'append', chunksize = 100000)
ontime_2004.to_sql('ontime', con = conn, index = False, if_exists = 'append', chunksize = 100000)
ontime_2005.to_sql('ontime', con = conn, index = False, if_exists = 'append', chunksize = 100000)
airports.to_sql('airports', con = conn, index = False)
carriers.to_sql('carriers', con = conn, index = False)
planes.to_sql('planes', con = conn, index = False)


c = conn.cursor()


#Question 2
#Which of the following airplanes has the 
#lowest associated average departure delay (excluding cancelled and diverted flights)?

               
q2 = c.execute('''
               SELECT model, avg(DepDelay) as average_delay
               FROM ontime JOIN planes USING (TailNum)
               WHERE ontime.diverted = 0 AND ontime.cancelled = 0 AND model IN ('737-230', 'ERJ 190-100 IGW', 'A330-223', '737-282')
               GROUP BY model
               ORDER BY average_delay
''').fetchall()

pd.DataFrame(q2)

#answer: 737-282

#-------

#Question 3
#Which of the following cities has the highest number of inbound flights
#excluding cancelled flights

q3 = c.execute('''
               SELECT airports.city AS city, COUNT(*) AS count_flights
               FROM ontime JOIN airports ON ontime.dest = airports.iata
               WHERE ontime.cancelled = 0 AND city IN ('Chicago', 'Atlanta', 'New York', 'Houston')
               GROUP BY city
               ORDER BY count_flights DESC
''').fetchall()

pd.DataFrame(q3)

#answer: Chicago

#-------

#Question 4
#Which of the following companies has the highest number of cancelled flights?

q4 = c.execute('''
               SELECT Description, sum(cancelled) as number_cancelled_flights
               FROM ontime JOIN carriers ON ontime.UniqueCarrier = carriers.Code
               WHERE Description IN ('United Air Lines Inc.', 'American Airlines Inc.', 'Pinnacle Airlines Inc.', 'Delta Air Lines Inc.')
               GROUP BY Description
               ORDER BY number_cancelled_flights DESC
''').fetchall()

pd.DataFrame(q4)

#answer: Delta Air Lines Inc.

#-------

#Question 5
#Which of the following companies has the highest number of cancelled flights,
#relative to their number of total flights?

q5 = c.execute('''
               SELECT Description, (sum(cancelled)*1.0/count(*)) as cancellation_rate
               FROM ontime JOIN carriers ON ontime.UniqueCarrier = carriers.Code
               WHERE Description IN ('United Air Lines Inc.', 'American Airlines Inc.', 'Pinnacle Airlines Inc.', 'Delta Air Lines Inc.')
               GROUP BY Description
               HAVING count(*)>0
               ORDER BY cancellation_rate DESC
''').fetchall()

pd.DataFrame(q5)

#answer: United Air Lines Inc

#answers in csv
q2_df = pd.DataFrame(q2)
q2_df.to_csv('q2.csv', index=False)
q3_df = pd.DataFrame(q3)
q3_df.to_csv('q3.csv', index=False)
q4_df = pd.DataFrame(q4)
q4_df.to_csv('q4.csv', index=False)
q5_df = pd.DataFrame(q5)
q5_df.to_csv('q5.csv', index=False)


#close everything
c.close()
conn.close()
